docs/adr/0006-filters.md

```diff
@@ -0,0 +1,253 @@
+- pilot the query parameter validation (current QueryParameterValidator bases itself on the given documentation schema) this is good but lacks flexibility when you need custom validation (created by @jdeniau)
+- compose with filters, which will naturally help creating an or/and filter
+- remove the relation between a query parameter and a property (they may have different names [#5980][pull/5980]), different types, a query parameter can have no link with a property (order filter)
+- provide a way to implement different query parameter syntaxes without changing the Filter implementation behind it
+
+We will keep a BC layer with the current doctrine system as it shouldn't change much.
+
+### Filter composition
+
+For this to work, we need to consider a 4 year old bug on searching with UIDs. Our SearchFilter allows to search by `propertyName` or by relation, using either a scalar or an IRI: 
+
+```
+/books?author.id=1
+/books?author.id=/author/1
+```
+
+Many attempts to fix these behavior on API Platform have lead to bugs and to be reverted. My proposal is to change how filters are applied to provide filters with less logic, that are easier to maintain and that do one thing good. 
+
+For the following example we will use an UUID to represent the stored identifier of an Author resource.
+
+We know `author` is a property of `Book`, that represents a Resource. So it can be filtered by:
+
+- IRI
+- uid
+
+We should therefore call both of these filters for each query parameter matched: 
+
+- IriFilter (will do nothing if the value is not an IRI)
+- UuidFilter
+
+With that in mind, an `or` filter would call a bunch of filters specifying the logic operation to execute. 
+
+### Query parameter
+
+The above shows that a query parameter **key**, which is a `string` may lead to multiple filters being called. This same can represent one or multiple values, and for a same **key** we can handle multiple types of data.
+Also, if someone wants to implement the [loopback API](https://loopback.io/doc/en/lb2/Fields-filter.html) `?filter[fields][vin]=false` the link between the query parameter, the filter and the value gets more complex. 
+
+We need a way to instruct the program to parse query parameters and produce a link between filters, values and some context (property, logical operation, type etc.). The same system could be used to determine the **type** a **filter** must have to pilot query parameter validation and the JSON Schema. 
+
+## Considered Options
+
+Let's define a new Attribute `Parameter` that holds informations (filters, context, schema) tight to a parameter `key`. 
+
+```php
+namespace ApiPlatform\Metadata;
+
+use ApiPlatform\OpenApi;
+
+final class Parameter {
+    public string $key;
+    public \ArrayObject schema;
+    public array $context;
+    public OpenApi\Parameter $openApi;
+    public string|callable provider(): Operation;
+    // filter service id
+    public string $filter;
```

Comment:
Maybe could we allow inline filter definitions?

```suggestion
    public string|callable $filter;
```

Reply:
I thought about that but with which interface? at first I wanted something like `mixed $value, Parameter $parameter, array $context` but it doesn't ally with `FilterInterface`